File: nQueens.cpp
#include<iostream> 
#include<ctime>
#include<cstdio>
using namespace std; 

const int MAXN = 50; 

int N; 
int avail[MAXN][MAXN];

void forward_checking(int n, int a) {
	for (int i = n; i < N; ++i) {
		avail[i][a] ++;
		if (a + i - n < N)
			avail[i][a + i - n] ++;
		if (a - i + n >= 0)
			avail[i][a - i + n] ++;
	}
}

void reset_forward_checking(int n, int a) {
	for (int i = n; i < N; ++i) {
		avail[i][a] --;
		if (a + i - n < N)
			avail[i][a + i - n] --;
		if (a - i + n >= 0)
			avail[i][a - i + n] --;
	}
}

int dfs_forward_checking(int n) {
	if (n >= N) {
		return 1;
	}
	int ret = 0;
	for (int i = 0; i < N; ++i) if (avail[n][i] == 0) {
		forward_checking(n, i);
		ret += dfs_forward_checking(n + 1);
		reset_forward_checking(n, i);
	}
	return ret;
}

int nQueens(int n) {
	N = n;
	for (int i = 0; i < n; ++i)
		for (int j = 0; j < n; ++j)
			avail[i][j] = 0;
	return dfs_forward_checking(0);
}

//=======================================================

int state[MAXN];

bool check(int n, int a) {
	for (int i = 0; i < n; ++i) {
		if (state[i] == a || 
			state[i] == a + n - i || 
			state[i] == a - n + i)
			return false;
	}
	return true;
}

int dfs_backtracking(int n) {
	if (n >= N) {
		return 1;
	}
	int ret = 0;
	for (int i = 0; i < N; ++i) if (check(n, i)) {
		state[n] = i;
		ret += dfs_backtracking(n + 1);
	}
	return ret;
}


int nQueens2(int n) {
	N = n;
	return dfs_backtracking(0);
}



int main() {
	freopen("nQueens.txt", "w", stdout);

	int n;
	// cout << "Please input n: ";
	cin >> n;

	for (int i = 1; i <= n; ++i) {
		auto start = clock();
		int ans = nQueens(i);
		auto end = clock();
	
		// cout << i << " " << ans << endl;
		cout << double(end - start) / CLOCKS_PER_SEC << " ";

		start = clock();
		ans = nQueens2(i);
		end = clock();
	
		// cout << i << " " << ans << endl;
		cout << double(end - start) / CLOCKS_PER_SEC << endl;
	}
	
	return 0;
}

File: nqueens.hs

import Data.List

judge :: Int -> Int -> [Int] -> Int -> Bool
judge maxn n state a = all f (zip [(n-1), (n-2)..1] state)
    where f (i, ai) = ai /= a && ai /= a + n - i && ai /= a - n + i

back_tracking :: Int -> Int -> [Int] -> Int
back_tracking maxn n state = if n > maxn then 1 else
    let avail = filter (judge maxn n state) [1..maxn] in 
        sum [back_tracking maxn (n + 1) (a:state) | a <- avail]

forwarding :: Int -> Int -> Int -> [[Int]] -> [[Int]]
forwarding maxn n a ass = [filter (g i) as | (i, as) <- zip [(n+1)..maxn] ass]
    where 
        g i ai = ai /= a && ai /= a + n - i && ai /= a - n + i
        f (i, as) = filter (g i) as

forward_checking :: Int -> Int -> [[Int]] -> Int
forward_checking _ _ [] = 1
forward_checking maxn n (as:ass) = 
    sum [forward_checking maxn (n+1) (forwarding maxn n a ass) | a <- as]

nQueens :: Int -> Int 
nQueens n = forward_checking n 1 $ replicate n [1..n]